test/e2e/storage/external/external.go

```diff
@@ -0,0 +1,214 @@
+	Describe(description, func() {
+		testsuites.DefineTestSuite(driver, csiTestSuites)
+	})
+
+	return nil
+}
+
+func (t testDriverParameter) loadDriverDefinition(filename string) (*driverDefinition, error) {
+	if filename == "" {
+		return nil, errors.New("missing file name")
+	}
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	// Some reasonable defaults follow.
+	driver := &driverDefinition{
+		DriverInfo: testsuites.DriverInfo{
+			SupportedFsType: sets.NewString(
+				"", // Default fsType
+			),
+		},
+		ClaimSize: "5Gi",
+	}
+	// TODO: strict checking of the file content once https://github.com/kubernetes/kubernetes/pull/71589
+	// or something similar is merged.
+	if err := runtime.DecodeInto(legacyscheme.Codecs.UniversalDecoder(), data, driver); err != nil {
+		return nil, errors.Wrap(err, filename)
+	}
+	return driver, nil
+}
+
+var _ testsuites.TestDriver = &driverDefinition{}
+
+// We have to implement the interface because dynamic PV may or may
+// not be supported. driverDefinition.SkipUnsupportedTest checks that
+// based on the actual driver definition.
+var _ testsuites.DynamicPVTestDriver = &driverDefinition{}
+
+// Needed only for deserialization which isn't done, therefore the
+// functions below can be empty stubs.
+var _ runtime.Object = &driverDefinition{}
+
+// DriverDefinition needs to be filled in via a .yaml or .json
+// file. It's methods then implement the TestDriver interface, using
+// nothing but the information in this struct.
+type driverDefinition struct {
+	// DriverInfo is the static information that the storage testsuite
+	// expects from a test driver. See test/e2e/storage/testsuites/testdriver.go
+	// for details. The only field with a non-zero default is the list of
+	// supported file systems (SupportedFsType): it is set so that tests using
+	// the default file system are enabled.
+	DriverInfo testsuites.DriverInfo
+
+	// ShortName is used to create unique names for test cases and test resources.
+	ShortName string
+
+	// StorageClass is required for enabling dynamic provisioning tests.
+	// When set to "default", a StorageClass with DriverInfo.Name as provisioner is used.
+	// Everything else is treated like the filename of a .yaml or .json file
+	// that defines a StorageClass.
+	StorageClass string
+
+	// ClaimSize defines the desired size of dynamically
+	// provisioned volumes. Default is "5GiB".
+	ClaimSize string
+
+	// ClientNodeName selects a specific node for scheduling test pods.
+	// Can be left empty.
+	ClientNodeName string
+}
+
+func (d *driverDefinition) DeepCopyObject() runtime.Object {
+	return nil
+}
+
+func (d *driverDefinition) GetObjectKind() schema.ObjectKind {
+	return nil
+}
+
+func (d *driverDefinition) GetDriverInfo() *testsuites.DriverInfo {
+	return &d.DriverInfo
+}
+
+func (d *driverDefinition) SkipUnsupportedTest(pattern testpatterns.TestPattern) {
+	supported := false
+	// TODO (?): add support for more volume types
```

Comment:
I think the volume types supported should also be defined by the external driver. Some may support `Block` or `Snapshots` in the future and you'd probably want to be able to toggle those on/off through the `DriverInfo` definition or some other method.

Reply:
This isn't about snapshot or blocks. Other `pattern.VolType` values are currently:
- PreprovisionedPV
- InlineVolume

I think those cannot be tested with a generic test driver.

https://github.com/kubernetes/kubernetes/blob/076af3d1b5cc1c368dd00f64ceefda8a87c1dfc7/test/e2e/storage/testpatterns/testpattern.go#L36-L46

But snapshots is something that can be added. It's not in the PR yet because it was added to the `TestDriver` interface after my initial version of this PR.

I think block device don't need any special support. It's a capability, and that can be set already.

Reply:
I've added snapshotting support. I've tried it manually and the test ran, but failed because my cluster lacked some feature gate:
```
• Failure in Spec Teardown (AfterEach) [365.125 seconds]
External Storage [Driver: csi-hostpath]
/nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/external/external.go:78
  [Testpattern: Dynamic PV (default fs)] provisioning
  /nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/testsuites/base.go:75
    should provision storage with snapshot data source [Feature:VolumeSnapshotDataSource] [AfterEach]
    /nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/testsuites/provisioning.go:202

    Expected error:
        <*errors.StatusError | 0xc0005290e0>: {
            ErrStatus: {
                TypeMeta: {Kind: "", APIVersion: ""},
                ListMeta: {SelfLink: "", ResourceVersion: "", Continue: ""},
                Status: "Failure",
                Message: "PersistentVolumeClaim \"pvc-6xw22\" is invalid: spec.dataSource: Forbidden: VolumeSnapshotDataSource is disabled by feature-gate",
                Reason: "Invalid",
                Details: {
                    Name: "pvc-6xw22",
                    Group: "",
                    Kind: "PersistentVolumeClaim",
                    UID: "",
                    Causes: [
                        {
                            Type: "FieldValueForbidden",
                            Message: "Forbidden: VolumeSnapshotDataSource is disabled by feature-gate",
                            Field: "spec.dataSource",
                        },
                    ],
                    RetryAfterSeconds: 0,
                },
                Code: 422,
            },
        }
        PersistentVolumeClaim "pvc-6xw22" is invalid: spec.dataSource: Forbidden: VolumeSnapshotDataSource is disabled by feature-gate
    not to have occurred
```

Reply:
Why not PreprovisionedPV? Like StorageClass, there could be a PersistentVolume field with FromName or FromFile and its source could be returned in GetPersistentVolumeSource. That said, other than maybe for the nfs and iscsi csi drivers IDK if it will be useful

Reply:
It might be useful for CSI drivers in progress. e.g. I haven't implemented CreateVolume yet but want to make sure publish works.